### ISSUE TITLE
No chat history

### ISSUE THREAD

**Issue by signalguest** [cryptocat/cryptocat#73]
Is that a bug?

**Comment by kaepora** [comment]
Not a bug, more like a missing feature that I'm not sure would fit in with Cryptocat's goal as privacy-centric chat software.

**Comment by ovalseven8** [comment]
If the chat history will be saved encrypted, I don't see any reason why this feature should not be added. Of course, opt-in/opt-out.

**Comment by bungabunga** [comment]
i'd also love to have a chat history. i think most user's threat model is not that tight. anyways, it could always be turned of if/when needed.

**Comment by tuxication** [comment]
I'm not a huge fan of complete chat history but would like to have maybe at least 2-3 past messages available. It is sometime hard to understand one message out of context received later when you get online.

**Comment by the-solipsist** [comment]
Storing chat histories (even in an encrypted form) defeats the purpose of forward secrecy.  That said, I'm sure many user would like to have the choice of (encrypted) chat histories.

**Comment by kaepora** [comment]
I've thought about this and right now my goal is to (when I have time):

* Implement chat history, turned off by default. Can be turned on globally for the account but not on a chat-by-chat basis because I think the latter makes the UX too confusing.
* Chat history will save last 100 messages, nothing indefinite.
* Nothing will be encrypted. Encrypting chat history means having a "chat history encryption password" -- and I insist on not making Cryptocat users have any password other than their login password.

No matter how much you argue, I won't make them have another password for any other part of the software, sorry. This is the same reason why your local account settings and identity keys are not encrypted. You might ask "why not encrypt with the account password?" -- that doesn't work because if the user changes their account password on another device, the data at rest on the rest of their devices will still be encrypted under their old password, hence also creating two passwords.

**Comment by TumM0reNcE5SsIti** [comment]
I agree with @the-solipsist that implementing chat histories LOCALLY defeats the purpose of forward secrecy. However, if we maintain these histories server-side without any content other than text, I can see this being a huge advancement in the Cryptocat architecture. If it is implemented in a way that 100 messages, server side, are encrypted using the account information we already have to store offline messages, then our messages are secure, and we are provided with a way to see last messages.

Server-side encryption is the most viable option in my opinion, not locally encrypted containers.
surespot is a good example of this implementation.

**Comment by IpsmLorem** [comment]
I agree with @kaepora, this feature must be turned off by default

**Comment by Gaboose** [comment]
@kaepora I don't understand the multiple passwords issue. Why does encrypting local data (either history or users.db) requires more than one user password? Can't the same one be used?

**Comment by gdebaisi** [comment]
Hello,

Does anyone know if this feature will be implemented soon ?

**Comment by andersbateva** [comment]
@kaepora , logging 100 unencrypted messages by the choice of only one of the buddies wouldn't be unfair with the other(s)? I mean, and what if Buddy A don't want to be logged in Buddy B's machine? What if Buddies A, C, D, E, F, G, H, J, K, and L don't want to be logged, but Buddy B decided to log 100 messages from each of them, resulting in 1000 messages being stored unencrypted and against the will of each of these?

**Comment by TumM0reNcE5SsIti** [comment]
@andersbateva, technically, all I have to do is copy paste, and BAM, logged (regardless of data-at-rest medium I choose). It's a neat idea to have an implementation, but I agree, it isn't necessary from a server-side perspective. I withdraw my interest. Always-on OTR is in our best interest.

**Comment by andersbateva** [comment]
@TumM0reNcE5SsIti , I highly doubt people would manually copy, paste, and BAM 100 messages, for each contact, each time they exchanged 100 messages. But doing it automatically will guarantee that it's happening.
